torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3004,6 +3005,164 @@ struct to_ir {
+            if (type_hint && type != type_hint) {
+              throw ErrorReport(apply.range())
+                  << "Expected " << type->repr_str()
+                  << " type annotation for dict(), found "
+                  << type_hint->repr_str();
+            }
+          }
+          /// Otherwise, use the default of Dict[str, Tensor]. If we have an
+          // iterable in apply.inputs(), then this default type will be
+          // overridden when the iterable is emitted
+          else {
+            type = DictType::create(StringType::get(), TensorType::get());
+          }
+        }
+
+        // If we haven't been passsed an iterable/mapping, we have
+        // either an empty call to `dict()` or we have kwargs
+        if (apply.inputs().empty()) {
+          std::vector<Value*> keys;
+          std::vector<Value*> values;
+          for (auto attr : apply.attributes()) {
+            auto sl = StringLiteral::create(apply.range(), attr.name().name());
+            keys.push_back(emitStringLiteral(sl));
+            values.push_back(emitExpr(attr.value()));
+          }
+          return std::make_shared<SimpleValue>(
+              graph
+                  ->insertNode(graph->createDict(
+                      type->expect<DictType>()->getKeyType(),
+                      type->expect<DictType>()->getValueType(),
+                      keys,
+                      values))
+                  ->output());
+        }
+
+        // Otherwise, we need to emit a new dict using the
+        // iterable/mapping we've been passed (and potentially kwargs
+        // as well)
+        else {
+          // We can't feasibly register all possible key x value
+          // combinations of new prim ops for the case that we use the
+          // constructor with a dict literal. It makes much more sense
+          // to transform the dict literal into a list of tuples so that
+          // we can use the existing constructors
+          if (!apply.inputs().empty() &&
+              apply.inputs()[0].kind() == TK_DICT_LITERAL) {
+            auto dict_lit = DictLiteral(apply.inputs()[0]);
+            std::vector<Expr> zipped;
+            for (auto key_it = dict_lit.key_inputs().begin(),
+                      val_it = dict_lit.value_inputs().begin();
+                 key_it != dict_lit.key_inputs().end();
+                 ++key_it, ++val_it) {
+              auto tuple_inputs =
+                  List<Expr>::create(apply.range(), {*key_it, *val_it});
+              auto tuple = TupleLiteral::create(apply.range(), tuple_inputs);
+              zipped.push_back(tuple);
+            }
+            auto ll_values = List<Expr>::create(apply.range(), zipped);
+            auto ll = ListLiteral::create(apply.range(), ll_values);
+            auto expr_list = List<Expr>::create(apply.range(), {ll});
+            // Change `apply` to a new Apply node holding a list of
+            // tuples
+            apply = Apply::create(
```

Comment:
Reusing `apply` here is tricky. Could we have a test case to ensure `apply.attributes()` is carried over?

Something like this:
>>> a = dict({"key1":2, "key2":4}, key3=3)
>>> a
{'key1': 2, 'key2': 4, 'key': 3}

Reply:
I think this is the same test as `test_dict_keyword_with_mixed_mapping_and_kwargs`. Please let me know if there's something that test is missing